DVR/URLRequest.swift

```diff
@@ -1,74 +1,85 @@
-extension NSURLRequest {
+extension URLRequest {
     var dictionary: [String: AnyObject] {
         var dictionary = [String: AnyObject]()
 
-        if let method = HTTPMethod {
-            dictionary["method"] = method
+        if let method = httpMethod {
+            dictionary["method"] = method as AnyObject?
         }
 
-        if let url = URL?.absoluteString {
-            dictionary["url"] = url
+        if let url = url?.absoluteString {
+            dictionary["url"] = url as AnyObject?
         }
 
         if let headers = allHTTPHeaderFields {
-            dictionary["headers"] = headers
+            dictionary["headers"] = headers as AnyObject?
         }
 
-        if let data = HTTPBody, body = Interaction.encodeBody(data, headers: allHTTPHeaderFields) {
-            dictionary["body"] = body
+        if let data = httpBody, let body = Interaction.encodeBody(data, headers: allHTTPHeaderFields) {
+            dictionary["body"] = body as AnyObject?
         }
 
         return dictionary
     }
 }
 
 
-extension NSURLRequest {
-    func requestByAppendingHeaders(headers: [NSObject: AnyObject]) -> NSURLRequest {
-        let request = mutableCopy() as! NSMutableURLRequest
+extension URLRequest {
+    func requestByAppendingHeaders(_ headers: [AnyHashable: Any]) -> URLRequest {
+        var request = self
         request.appendHeaders(headers)
-        return request.copy() as! NSURLRequest
+        return request
     }
 
-    func requestWithBody(body: NSData) -> NSURLRequest {
-        let request = mutableCopy() as! NSMutableURLRequest
-        request.HTTPBody = body
-        return request.copy() as! NSURLRequest
+    func requestWithBody(_ body: Data) -> URLRequest {
+        var request = self
+        request.httpBody = body
+        return request
     }
 }
 
 
-extension NSMutableURLRequest {
-    convenience init(dictionary: [String: AnyObject]) {
-        self.init()
+extension URLRequest {
+    init?(dictionary: [String: AnyObject]) {
 
-        if let method = dictionary["method"] as? String {
-            HTTPMethod = method
+        guard let method = dictionary["method"] else {
+            print("No method in \(dictionary)")
+            return nil
         }
 
-        if let string = dictionary["url"] as? String, url = NSURL(string: string) {
-            URL = url
+        guard let string = dictionary["url"] as? String, let furl = URL(string: string) else {
+            print("No url in \(dictionary)")
+            return nil
         }
 
-        if let headers = dictionary["headers"] as? [String: String] {
-            allHTTPHeaderFields = headers
+        var headers: [String: String] = [:]
+        if let h = dictionary["headers"] as? [String: String] {
+            headers = h
         }
 
-        if let body = dictionary["body"] {
-            HTTPBody = Interaction.dencodeBody(body, headers: allHTTPHeaderFields)
+        var body: Data? = nil
+
+        if let b = dictionary["body"] as? String {
```

Comment:
This will always fail. No decode is ever run and "body" key is never a string. Consider instead:

        var b: Data? = nil

        if let body = dictionary["body"] {
            b = Interaction.decodeBody(body, headers: headers)
        }